### PR TITLE
Patterns: Add Category Not Found page

### PR DESCRIPTION
**Pull request by stephanethomas**
This pull request addresses https://github.com/Automattic/dotcom-forge/issues/5868 by adding a new `404` page for the `Patterns` section. This error page is shown whenever a user navigates to a url which uses an invalid category:

image

Finally, this pull request also splits the `controller.tsx` file into `constants.ts` and `paths.ts` files to better separate concerns, and use names that more accurately convey what they contain. It also cleans up and refines the `404` pages from the `Themes` and `Plugins` sections, and fixes their call-to-action to retain the locale.

#### Additional notes

My original goal was to build a page that:

* Would render server-side
* Would return a `404` HTTP status code
* Would render the header and footer
* Would be displayed when query parameters are shown

Unfortunately, I wasn't able to achieve all of that. I went through several iterations, trying different approaches ([for example](https://github.com/Automattic/wp-calypso/pull/88312/commits/047df2a8252444e5300fc6e743070974c4de85ff)). Testing of server-side rendering has been a slow process because Calypso needs to be restarted for every change. Caching also plays an important role, as HTTP statuses and context parameters are never cached. Moreover, we cache the markup early, which means the code that would check the category and return a `404` will only run on the first request, not subsequent ones (a `200` would then be returned). In the end, I lost a non negligible amount of time, and had to make compromises.

#### Testing instructions

1. Run `git checkout add/patterns-category-not-found-page` and start your server, or access a [live branch](https://github.com/Automattic/wp-calypso/pull/88312#issuecomment-1984051354)
2. Make sure you are logged out
3. Assert that the new error page is shown for the following urls:
   * http://calypso.localhost:3000/patterns/something
   * http://calypso.localhost:3000/fr/patterns/something
   * http://calypso.localhost:3000/theme/something
   * http://calypso.localhost:3000/es/theme/something
   * http://calypso.localhost:3000/plugins/something
   * http://calypso.localhost:3000/fr/plugins/something
4. Assert that a `404` HTTP status is returned every time your reload a `Patterns` page
5. Assert that the call-to-action point to the same locale
6. Test again all those steps while logged-in

### CODE REVIEW

File: client/my-sites/patterns/wrapper.tsx [hunk 1]
```diff
@@ -14,9 +14,7 @@ export const PatternsWrapper = ( { children }: PatternsWrapperProps ) => {
 		<>
 			{ isLoggedIn && <UniversalNavbarHeader isLoggedIn /> }
 
-			<Main isLoggedOut fullWidthLayout>
-				{ children }
-			</Main>
+			<Main fullWidthLayout>{ children }</Main>
```
Comment (stephanethomas) [inline client/my-sites/patterns/wrapper.tsx:17]: I'm removing this prop, we don't use it.

File: client/my-sites/patterns/components/category-not-found/style.scss [hunk 2]
```diff
@@ -0,0 +1,6 @@
+.is-section-patterns {
+	.empty-content {
+		margin-bottom: 60px;
+		margin-top: 60px;
+	}
```
Comment (fredrikekelund) [inline client/my-sites/patterns/components/category-not-found/style.scss:10]:
```suggestion
		min-height: 50vh;
	}
```
It'd be nice to make this slightly taller by default. It looks a bit cramped right now, IMO.

File: client/my-sites/patterns/components/category-not-found/index.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+export const PatternsCategoryNotFound = () => {
+	const isLoggedIn = useSelector( isUserLoggedIn );
+	const translate_not_yet = useTranslate();
+
+	return (
+		<EmptyContent
+			title={ translate_not_yet( "Oops! We can't find this category!" ) }
+			line={ translate_not_yet( "The category you are looking for doesn't exist." ) }
+			action={ translate_not_yet( 'Browse all patterns' ) }
```
Comment (fredrikekelund) [inline client/my-sites/patterns/components/category-not-found/index.tsx]: Should we add comments to these strings..?

**Review by fredrikekelund — APPROVED** [review]
LGTM 👍 I had two small suggestions, but beyond that, this works as expected.

http://calypso.localhost:3000/plugins/something doesn't return a 404 status, but that appears to be the same behavior on wordpress.com right now. Feel free to add this to our backlog, but let's not spend anymore time on this right now, since it's not part of the pattern library effort.

File: client/my-sites/patterns/components/category-not-found/style.scss (hunk 2 above)
Comment (stephanethomas) [reply client/my-sites/patterns/components/category-not-found/style.scss:10]: That's because the `margin-top` rule is overridden. Anyway, I'll fix that 🙂.

**Pull request merged**